### PR TITLE
Feat/encryption system rework with password

### PR DESCRIPTION
**Pull request by MasterKenth**
## Explain the changes you’ve made
* Added a new encryption system that works with abstract "strategies" for encryption and decryption.
  * Two strategies are included; DeviceLocal (same as old "private AES" used for solo applicants), and Password-based (will replaced Symmetric Key)
* Deprecated existing encryption service and helper

Note: new encryption system is not being used by anything in this PR. Old encryption service is still being used.

## Explain why these changes are made

The current encryption system is not extensible enough and making changes are hard. This PR change was needed to facilitate implementing Password-based encryption ("pin code") as well as preparing to remove Symmetric Key exchange encryption.

## Explain your solution

Broadly two things have happened:

* The now deprecated `EncryptionService` and `EncryptionHelper` has been moved to `source/services/encryption/DEPRECATED` and files using them has been updated to point there (`CaseOverview` etc.).
* New files has been added to `source/services/encryption` that implements the new system, that are not used by anything yet.

### The new encryption system
Each method of encryption implements `IEncryptionStrategy`. They are oblivious to project-specific data structures (`Case`, `Form` etc.) and work only with raw data and abstract interfaces.

`CaseEncryptionService` is the glue between the generic strategies and the project-specific data. It handles encryption/decryption based on `Case` objects and `Form` objects. This is generally the class that will be called on by the UI layer or contexts.

`CaseEncryptionHelper` includes a set of commonly used types and functions as well as manages the mapping between `Case` encryption type and concrete encryption strategy (might move the second part to a separate concern?).

## How to test

Concrete example:

1. Checkout this branch
2. Make sure tests pass
3. Make sure app works as usual (start new case, encrypt ongoing, submit etc.)

## Tested environments

- [] Storybook on a iOS device/simulator.
- [x] Building the Application on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

## Other notes

---

## Checklist for reviewers

Here are a few points that are useful for reviewers to check.

The short list of the (maybe) most important points are:

- Does this PR work as expected?
- Does the PR properly only concern a single feature/bug/thing?
- Is the code clear and concise?
- Will the changes impact existing functionality?
- Are there any linting errors (ESLint) that should be resolved?
- Do the automatic GitHub checks pass?

However it is good if a reviewer can be as thorough as possible,
within reasonable limits of course.

### Functionality

- Does this PR work as expected?
- If this is a new feature, does it add value, or is it a sign of feature-creep?
- Will the changes impact existing functionality?
- Does the changes create inconsistencies somewhere else?
- Does the code work on different platforms if applicable?
- How does the code handle if unexpected values are passed to functions etc.?

### Readability, syntax, formatting, style

- Is the code clear and concise?
- Has styling rules (Prettier) been applied to included files?
- Are functions/variables/etc. aptly named?

### Design

- Is the code properly planned and designed?
- Is code organized/separated in a logical fashion?
- Does the code follow clean code practices to a reasonable degree?
- Is the code self-documenting enough?
- Is the code in sync with existing patterns/technologies?
- Does the code reuse existing functionality where applicable?
- Is the code designed with reusability in mind?
- Does the code have hard-coded values that can be refactored?

### Patterns, idioms, tests, best practices

- Are there any linting errors (ESLint) that should be resolved?
- Is the code properly typed (TypeScript)?
- Does the code utilize language features to their full potential?

### Debugging, testing, configuring

- Is appropriate logging used?
- Have tests been added/modified to accommodate the changes?
- Is the feature sufficiently configurable in case tweaks are required in the future?

### Performance, reliability, scalability

- Is the code well optimized?
- Does the code scale?
- Does the code handle failure scenarios?

### Security

- Is there any obvious potential for exploits and security threats?
- Does code check proper authorization where applicable?

### Etiquettes

- Are commits atomic (does only one thing per commit)?
- Does the PR properly only concern a single feature/bug/thing?
- Are commit messages well-written?
- Do the automatic GitHub checks pass?

### CODE REVIEW

**Comment by helsingborgsstad** [comment]
Task linked: [CU-21w373e Implementera kryptering med lösenord/pinkod](https://app.clickup.com/t/21w373e)

File: source/services/encryption/DeviceLocalAESStrategy.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+export interface DeviceLocalAESParams {
+  key: string;
+  iv: string;
+}
+
+async function createAesParams(): Promise<DeviceLocalAESParams> {
+  const randomBytes = await Aes.randomKey(16);
+
+  const key = await Aes.pbkdf2(randomBytes, "salt4D42bf960Sm1", 5000, 256);
```
Comment (karlknowit) [inline source/services/encryption/DeviceLocalAESStrategy.ts:19]: I suggest that values like `"salt4D42bf960Sm1", 5000, 256` are put in constants for readability

File: source/services/encryption/PasswordStrategy.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+const { Aes } = NativeModules;
+
+export interface PasswordParams {
+  password: string;
+}
+
+async function generateAESParamsFromPassword(
+  password: string
+): Promise<DeviceLocalAESParams> {
+  const key = await Aes.pbkdf2(password, "salt4D42bf960Sm1", 5000, 256);
```
Comment (karlknowit) [inline source/services/encryption/PasswordStrategy.ts:21]: As mentioned above, put in constants. These values are also in multiple places which makes it even better to have those in constants.

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+  PasswordParams,
+  PasswordStrategy,
+} from "../PasswordStrategy";
+import { to } from "../../../helpers/Misc";
+import {
+  EncryptionContext,
+  EncryptionDependencies,
+} from "../EncryptionStrategy";
+
+function MakeMockCase(
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:49]: MakeMockCase -> makeMockCase

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+const MOCK_GET_USER = () =>
+  Promise.resolve<UserInterface>({ personalNumber: "198602282389" });
+
+function makeEncryptionService(storage: IStorage): ICaseEncryptionService {
+  return new CaseEncryptionService(storage, MOCK_GET_USER);
+}
+
+describe("CaseEncryptionService (CaseEncryptionHelper)", () => {
+  test("getCurrentForm", () => {
+    const form = getCurrentForm(CASE_DECRYPTED_SOLO);
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:139]: missing newline before expect in some tests

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+    await PasswordStrategy.generateAndSaveBasicPinPassword(
+      context,
+      dependencies
+    );
+    const secondCheck = await PasswordStrategy.hasPassword(
+      context,
+      dependencies
+    );
+
+    expect(firstCheck).toEqual(false);
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:681]: `toEqual` -> `toBe` for primitive values

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+      {
+        encryptionDetails: getEncryptionFromCase(CASE_DECRYPTED_PARTNER),
+      },
+      { storage: mockStorage }
+    );
+    const newCase = await encryptionService.encrypt(CASE_DECRYPTED_PARTNER);
+    const { answers, encryption } = getCurrentForm(newCase);
+
+    expect(answersAreEncrypted(answers)).toBe(true);
+    expect(encryption.type).toEqual(EncryptionType.PASSWORD);
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:634]: toEqual -> toBe

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+    };
+
+    const password = await PasswordStrategy.generateAndSaveBasicPinPassword(
+      {
+        encryptionDetails: getEncryptionFromCase(CASE_DECRYPTED_PARTNER),
+      },
+      { storage: mockStorage }
+    );
+
+    expect(typeof password).toEqual("string");
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:612]: toEqual -> toBe

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+    const context: EncryptionContext = {
+      encryptionDetails: getEncryptionFromCase(CASE_DECRYPTED_PARTNER),
+    };
+    const dependencies: EncryptionDependencies = { storage: mockStorage };
+    const testPassword = "test password";
+
+    await PasswordStrategy.providePassword(testPassword, context, dependencies);
+    const password = await PasswordStrategy.getPassword(context, dependencies);
+
+    expect(password).toEqual(testPassword);
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:696]: toEqual -> toBe

File: source/services/encryption/test/CaseEncryptionService.test.ts [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+      JSON.stringify(CASE_PARAMS_PARTNER)
+    );
+    const secondDecrypt = await encryptionService.decrypt(
+      CASE_ENCRYPTED_PARTNER
+    );
+
+    expect(firstDecrypt).toEqual(secondDecrypt);
+    expect(decryptError).toBeDefined();
+    expect(decryptError).toBeInstanceOf(EncryptionException);
+    expect((decryptError as EncryptionException).status).toEqual(
```
Comment (karlknowit) [inline source/services/encryption/test/CaseEncryptionService.test.ts:729]: toEqual -> toBe

**Review by karlknowit — CHANGES_REQUESTED** [review]

File: source/services/encryption/CaseEncryptionHelper.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  };
+
+  return makeFormWithNewAnswers(form, newAnswers, encryption);
+}
+
+export function makeFormWithDecryptedData(
+  form: AnsweredForm,
+  decryptedData: string
+): AnsweredForm {
+  const newAnswers: Answer[] = JSON.parse(decryptedData);
```
[the remark this answers is not in the document]
Comment (MasterKenth) [reply source/services/encryption/CaseEncryptionHelper.ts:179]: Good call; I've added a check and test for that.

File: source/services/encryption/CaseEncryptionHelper.ts [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+  answers: PossiblyEncryptedAnswers
+): answers is EncryptedAnswersWrapper {
+  return !Array.isArray(answers);
+}
+
+export function getEncryptionFromForm(form: AnsweredForm): EncryptionDetails {
+  return form.encryption;
+}
+
+export function getEncryptionFromCase(caseData: Case): EncryptionDetails {
```
[the remark this answers is not in the document]
Comment (MasterKenth) [reply source/services/encryption/CaseEncryptionHelper.ts:70]: In practice probably not, however the next few lines check for that so it should be ok (or at least possible to handle):

```typescript
  if (!caseData) {
    throw new EncryptionException(
      EncryptionErrorStatus.INVALID_CASE,
      "invalid case"
    );
  }
```

**Comment by sebastianthulin** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 #### 2 Issues 
* Anti-pattern 2 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/helsingborg-stad/repositories/fedfce3fd8695f238dea9cb00bd37ce7/pull-request/656)

**Review by karlknowit — APPROVED** [review]

**Pull request merged**